### PR TITLE
Move to new syntax

### PR DESCRIPTION
**Pull request by swertz**
Move the exporter to the current MoMEMta format.

Caveat: cmake & make file currently do not work, since the build is not standalone anymore but depends on MoMEMta itself... But it has to work without having to edit MoMEMta's CMakeLists file. I welcome any input there...

Small change needed in MoMEMta: move the initial partons' 4-momenta in the compute() call from a vector<> to a pair<> (since there are always two of them).

Todo: license/credits in generated files

### CODE REVIEW

**Comment by blinkseb** [comment]
Thanks a lot!
For the build, I think it's actually easy:
- No Makefile, it's a pain to maintain and it won't work anyway
- Treat MoMEMta like any external dependency, and use `find_package(MoMEMta)`. We'll need to add a small `FindMoMEMta.cmake` file, but it should be straightforward (just look for `momemta/momemta.h` and `libmomemta.so` in system path). It's basically here that you need to `make install` momemta if you want to be able to build external matrix elements.

**Comment by blinkseb** [comment]
Note that I don't expect it to work out of the box. There's for example the `logging.h` header with is not in public API while used in almost all the headers: you'll see a nice `File not found` error when building.

I'll try to make things in order this afternoon.

**Comment by swertz** [comment]
Thanks! I agree, we can remove the makefile.

There is no way to make it work without running `make install` for MoMEMta? If the matrix element is put in the `MatrixElements` folder of MoMEMta, there is no way to go look for CMakeLists files in subfolders of it, and build the matrix element when building MoMEMta itself but without modifying the "master" CMakeLists? (not very clear, sorry...)

**Comment by blinkseb** [comment]
> There is no way to make it work without running make install for MoMEMta? If the matrix element is put in the MatrixElements folder of MoMEMta, there is no way to go look for CMakeLists files in subfolders of it, and build the matrix element when building MoMEMta itself but without modifying the "master" CMakeLists? (not very clear, sorry...)

IMO it's two separate things:
- Either we talk about the matrix elements in `MatrixElements`
- Or we talk about _generic_ matrix elements anyone can export

The first item is a special case of 2), and you are right, there's no way to handle that automatically. But I don't think it's an issue, since this is something **we** do and we can update the cmake file as we add the new matrix element. Moreover, it's not the same output (either static library linked to `libmomemta.so` or directly a shared library). Maybe we can think of a `CMakeLists.txt` file which handle the two cases based on the value of a variable (declared when building MoMEMta and not when building in standalone)?

The user do not necessary need to do a `make install`, we need to provide a way to specify where MoMEMta is (something a la `BOOT_ROOT`, like `MOMEMTA_INCLUDEDIRS` and `MOMEMTA_LIBRARIES`). If it's installed system-wide, then the detection is automatic, otherwise you need to specify these variables.

I'm still in the process to clean the public headers, it's harder than expected because of the logging. It seems we'll have either to drop `spdlog` and switch to boost.log, or install `spdlog` with `MoMEMta` (not fan of this idea)...

**Comment by swertz** [comment]
> The user do not necessary need to do a make install, we need to provide a way to specify where MoMEMta is (something a la BOOT_ROOT, like MOMEMTA_INCLUDEDIRS and MOMEMTA_LIBRARIES). If it's installed system-wide, then the detection is automatic, otherwise you need to specify these variables.

That would be perfectly fine, and makes my comment about the `MatrixElements` folder in MoMEMta moot!

> It seems we'll have either to drop spdlog and switch to boost.log, or install spdlog with MoMEMta (not fan of this idea)...

Me neither. What's the problem with boost log, since we already link against boost (aside from the work needed to switch everywhere it's used)?

**Comment by swertz** [comment]
The output matrix elements now build correctly, thanks to @blinkseb's update to MoMEMta!

I didn't make a specific CMake module to find the package, as only the headers are needed here and that would clutter unneccessarily the generated folder (I leave that for the Tutorials).

It runs fine on ingrid, it'd be great if someone could check if it also works when MoMEMta is installed system-wide...

**Comment by blinkseb** [comment]
Very nice!! I'll try with a system-wide installation!

**Comment by blinkseb** [comment]
I'm wondering if the fact that we don't explictely link against `libmomemta.so` is a problem or not? Did you try to load the matrix element in `MoMEMta`?

And finally, did you try to change the `MOMEMTA_INCLUDE_DIRS` variable in the command line to see if it works?

**Comment by swertz** [comment]
> I'm wondering if the fact that we don't explictely link against `libmomemta.so` is a problem or not? Did you try to load the matrix element in `MoMEMta`?

Hm I haven't tried, you're right it should be done. I'll check!

> And finally, did you try to change the `MOMEMTA_INCLUDE_DIRS` variable in the command line to see if it works?

No I've used the `CMAKE_PREFIX_PATH` variable to configure the build. But I've checked and using `MOMEMTA_INCLUDE_DIR` (without S!) works as well!

**Comment by blinkseb** [comment]
> No I've used the CMAKE_PREFIX_PATH variable to configure the build. But I've checked and using MOMEMTA_INCLUDE_DIR (without S!) works as well!

Then it's perfect , thanks!

File: README.md [hunk 1]
```diff
@@ -1,2 +1,40 @@
...
+cd myHappyME
+mkdir build
+cd build
+cmake ..
+make -j 4
+```
+This generates a shared library that can be dynamically loaded by MoMEMta (using the `load_modules()` function in the Lua config file).
+
+The following options are available when configuring the the build (when running `cmake ..`):
+- `-DCMAKE_PREFIX_PATH=(path)`: Path to your installation of MoMEMta. Use this if your version of MoMEMta was installed locally and not in your system directories
```
Comment (blinkseb) [inline README.md]: I'd prefer to document the `MOMEMTA_INCLUDE_DIR` variable instead of tweaking `CMAKE_PREFIX_PATH`.

**Comment by swertz** [comment]
I've tried running an example loading the ME library dynamically: it works! :)

**Comment by blinkseb** [comment]
That's a **very good** news!! I'm still waiting for the `bzr branch` to finish and I'll try the exporter with a system wide installation :)

**Comment by blinkseb** [comment]
So:
- exporter works perfectly, I was able to export the ME without any issue (tried with `pp -> gamma gamma`)
- I was able to build the matrix element without issues with a system-wide installation of MoMEMta: the headers were correctly found
- I was able to load the matrix element in MoMEMta, and use it instead of the `tt` matrix element.

But:
- There's a crash at the end of the job in the library because of an access to an invalid memory location. Using valgrind, I see, as soon as the integration begin, a nasty `Invalid write of size 4` and then `Invalid write of size 8`.

I'll investigate, but this can't be merged for the moment.

**Comment by blinkseb** [comment]
Ok it seems to be because I simply switched the `tt` matrix element with the new one. The `tt` matrix element expects 6 particles, while the new one only two. No check are performed, so you write 6 particles in an array of two: crash.

**Comment by swertz** [comment]
Makes sense :D

File: Template/function_definitions.inc [hunk 2]
```diff
@@ -1,44 +1,41 @@
...
-	momenta[0] = (double*) (&initialMomenta[0][0]);
-    momenta[1] = (double*) (&initialMomenta[1][0]);
+	momenta[0] = (double*) (&initialMomenta.first[0]);
+    momenta[1] = (double*) (&initialMomenta.second[0]);
 
     // Suppose final particles are passed in the "correct" order
     std::vector<int> selectedFinalState(%(nexternal)s - 2);
     size_t index = 2;
-    for (auto const &finalPart: finalState){
+    for (auto const &finalPart: finalState) {
```
Comment (blinkseb) [inline Template/function_definitions.inc]: In order to fix https://github.com/MoMEMta/MoMEMta-MaGMEE/pull/3#issuecomment-219662965, I'd propose to change the `foreach` loop to an old style loop. We know here how many final particles we expect (`%(nexternal)s - 2` or `selectedFinalState.size()`), so we can use:

```
for (size_t index = 0; index < (%(nexternal)s - 2); index++) {
    selectedFinalState[index] = finalPart.first;
    momenta[index + 2] = (double*) (&finalPart.second[0]);
}
```

This at least would solve the _too much particles_

Comment (swertz) [inline Template/function_definitions.inc]: I agree, that's better.

File: Template/function_definitions.inc [hunk 3]
```diff
@@ -21,11 +21,10 @@ std::map< std::pair<int, int>, double > %(process_class)s::compute( const std::p
 
     // Suppose final particles are passed in the "correct" order
     std::vector<int> selectedFinalState(%(nexternal)s - 2);
-    size_t index = 2;
-    for (auto const &finalPart: finalState) {
-      selectedFinalState[index - 2] = finalPart.first;
-      momenta[index] = (double*) (&finalPart.second[0]);
-      index++;
+    for (size_t index = 0; index < (%(nexternal)s - 2); index++) {
```
Comment (blinkseb) [inline Template/function_definitions.inc:24]: You were a bit too prompt ;) I made some mistakes in my code snippet. The correct one is:

```
for (size_t index = 0; index < (%(nexternal)s - 2); index++) {
  selectedFinalState[index] = finalState[index].first; 
  momenta[index + 2] = (double * ) (&finalState[index].second[0]); 
}
```

(note: there's no `index++` no more)

Comment (blinkseb) [inline Template/function_definitions.inc:24]: You already did the right thing sorry misread the commit. There's just the `index++` to remove

**Comment by blinkseb** [comment]
I tested the `for loop` and it no longer crashed :) So big :+1: for me, thanks a lot for this work!

File: Template/function_definitions.inc (hunk 3 above)
Comment (swertz) [inline Template/function_definitions.inc:24]: Done!

**Comment by swertz** [comment]
Great, I'll rebase and push, I think for now this is ready to be merged!

**Comment by swertz** [comment]
Done!

**Comment by blinkseb** [comment]
Perfect, thanks a lot!

**Pull request merged**